Assignment 14

task 1

Create a contingency table in Excel or Google Sheets showing the relationship between users' favorite food delivery app (Zomato, Swiggy, Uber Eats) and their preferred payment method (Paytm, PhonePe, Credit Card) using a sample of at least 20 responses.

In [1]:
import pandas as pd

# Simulate 24 raw survey responses: each row is one user's favorite app and payment method
data = {
    "App": ["Swiggy", "Uber Eats", "Uber Eats", "Zomato", "Uber Eats", "Swiggy",
            "Zomato", "Swiggy", "Zomato", "Zomato", "Zomato", "Zomato",
            "Swiggy", "Swiggy", "Swiggy", "Swiggy", "Zomato", "Zomato",
            "Uber Eats", "Swiggy", "Uber Eats", "Uber Eats", "Uber Eats", "Swiggy"],
    "Payment Method": ["PhonePe", "PhonePe", "PhonePe", "Paytm", "Credit Card", "Paytm",
                        "Paytm", "PhonePe", "PhonePe", "Credit Card", "Paytm", "Paytm",
                        "PhonePe", "Paytm", "PhonePe", "Credit Card", "PhonePe", "Credit Card",
                        "Paytm", "Credit Card", "Credit Card", "Credit Card", "Paytm", "Credit Card"]
}

survey_df = pd.DataFrame(data)
survey_df.index = survey_df.index + 1
survey_df.index.name = "Response #"

print(f"Total responses: {len(survey_df)}\n")
print(survey_df)

# Build the contingency table using pandas crosstab, with row and column totals
contingency_table = pd.crosstab(survey_df["App"], survey_df["Payment Method"], margins=True, margins_name="Total")

print("\nContingency Table (App vs Payment Method):")
print(contingency_table)


Total responses: 24

                  App Payment Method
Response #                          
1              Swiggy        PhonePe
2           Uber Eats        PhonePe
3           Uber Eats        PhonePe
4              Zomato          Paytm
5           Uber Eats    Credit Card
6              Swiggy          Paytm
7              Zomato          Paytm
8              Swiggy        PhonePe
9              Zomato        PhonePe
10             Zomato    Credit Card
11             Zomato          Paytm
12             Zomato          Paytm
13             Swiggy        PhonePe
14             Swiggy          Paytm
15             Swiggy        PhonePe
16             Swiggy    Credit Card
17             Zomato        PhonePe
18             Zomato    Credit Card
19          Uber Eats          Paytm
20             Swiggy    Credit Card
21          Uber Eats    Credit Card
22          Uber Eats    Credit Card
23          Uber Eats          Paytm
24             Swiggy    Credit Card

Contingency Tabl

**Explanation:** We simulate 24 individual survey responses as a pandas DataFrame, with each row recording one user's favorite food delivery app and their preferred payment method. `pd.crosstab()` then builds the contingency table directly from this raw data, counting how many users fall into each App x Payment Method combination, along with row and column totals (`margins=True`). This is exactly what building the table manually in Excel or Google Sheets with COUNTIFS would produce, just done with pandas instead. The same counts from this table are reused in Task 3 below for the chi-square test of independence.

task 2

Given the following observed counts of music streaming preference among males and females: Spotify (Males: 12, Females: 18), Gaana (Males: 8, Females: 7), YouTube Music (Males: 10, Females: 15), calculate the expected counts for each cell assuming independence.

**Hint:** Use the formula: (row total x column total) / grand total.

In [2]:
import pandas as pd
import numpy as np

# Observed counts as a pandas DataFrame: rows = Gender, columns = App
observed_df = pd.DataFrame(
    {"Spotify": [12, 18], "Gaana": [8, 7], "YouTube Music": [10, 15]},
    index=["Male", "Female"]
)

print("Observed counts:")
print(observed_df)

row_totals = observed_df.sum(axis=1)
col_totals = observed_df.sum(axis=0)
grand_total = observed_df.values.sum()

print("\nRow totals:\n", row_totals)
print("\nColumn totals:\n", col_totals)
print(f"\nGrand total: {grand_total}")

# Expected count for each cell = (row total * column total) / grand total
expected_df = pd.DataFrame(
    np.outer(row_totals, col_totals) / grand_total,
    index=observed_df.index,
    columns=observed_df.columns
).round(2)

print("\nExpected counts (assuming independence):")
print(expected_df)


Observed counts:
        Spotify  Gaana  YouTube Music
Male         12      8             10
Female       18      7             15

Row totals:
 Male      30
Female    40
dtype: int64

Column totals:
 Spotify          30
Gaana            15
YouTube Music    25
dtype: int64

Grand total: 70

Expected counts (assuming independence):
        Spotify  Gaana  YouTube Music
Male      12.86   6.43          10.71
Female    17.14   8.57          14.29


**Explanation:** Under the assumption of independence (that gender and app preference are unrelated), the expected count for each cell is (row total x column total) / grand total. For example, expected Male-Spotify count = (30 x 30) / 70 ≈ 12.86, close to the observed 12. These expected counts form the baseline that a chi-square test compares the actual observed counts against.

task 3

Perform a chi-square test of independence using the data from your contingency table in task 1 to determine if there is a significant association between app choice and payment method. Show your calculation steps and state your conclusion.

In [3]:
from scipy.stats import chi2_contingency

# Reuse the contingency table counts from Task 1 (drop the margin totals row/column)
observed_counts = contingency_table.loc[["Zomato", "Swiggy", "Uber Eats"], ["Paytm", "PhonePe", "Credit Card"]]

print("Observed contingency table (from Task 1):")
print(observed_counts)

chi2_stat, p_value, dof, expected = chi2_contingency(observed_counts)

print(f"\nChi-square statistic = {round(chi2_stat, 4)}")
print(f"Degrees of freedom = {dof}")
print(f"p-value = {round(p_value, 4)}")

expected_df = pd.DataFrame(expected.round(2), index=observed_counts.index, columns=observed_counts.columns)
print("\nExpected counts (assuming independence):")
print(expected_df)

alpha = 0.05
if p_value < alpha:
    conclusion = "Reject H0 -> there IS a significant association between app choice and payment method"
else:
    conclusion = "Fail to reject H0 -> no significant association between app choice and payment method"

print(f"\nConclusion: {conclusion}")


Observed contingency table (from Task 1):
Payment Method  Paytm  PhonePe  Credit Card
App                                        
Zomato              4        2            2
Swiggy              2        4            3
Uber Eats           2        2            3

Chi-square statistic = 1.9524
Degrees of freedom = 4
p-value = 0.7445

Expected counts (assuming independence):
Payment Method  Paytm  PhonePe  Credit Card
App                                        
Zomato           2.67     2.67         2.67
Swiggy           3.00     3.00         3.00
Uber Eats        2.33     2.33         2.33

Conclusion: Fail to reject H0 -> no significant association between app choice and payment method


**Explanation:** `scipy.stats.chi2_contingency()` takes the observed contingency table directly from Task 1, automatically calculates the expected counts (using the same row-total x column-total / grand-total formula as Task 2), and computes the chi-square statistic as the sum of (observed - expected)^2 / expected across all cells, along with degrees of freedom ((rows-1) x (columns-1) = 2x2 = 4) and the p-value. With this data, the p-value comes out above 0.05, so we fail to reject the null hypothesis — meaning there isn't strong statistical evidence of a real association between which food delivery app someone prefers and which payment method they use.

task 4

For a new Flipkart product launch, suppose you surveyed 30 users and observed the following: 18 prefer the new product, 12 do not. If the company claims that 70% of users will prefer the new product, perform a chi-square goodness of fit test to check this claim.

**Hint:** Calculate expected counts based on the claimed proportion.

In [4]:
import numpy as np
from scipy.stats import chisquare

total_users = 30
observed_counts = np.array([18, 12])   # [prefer, do not prefer]

claimed_proportion_prefer = 0.70
claimed_proportion_not_prefer = 1 - claimed_proportion_prefer

expected_counts = np.array([
    total_users * claimed_proportion_prefer,
    total_users * claimed_proportion_not_prefer
])

print(f"Observed counts [Prefer, Do Not Prefer]: {observed_counts}")
print(f"Expected counts based on claim (70% prefer): {expected_counts}")

chi2_stat, p_value = chisquare(f_obs=observed_counts, f_exp=expected_counts)

print(f"\nChi-square statistic = {round(chi2_stat, 4)}")
print(f"p-value = {round(p_value, 4)}")

alpha = 0.05
if p_value < alpha:
    conclusion = "Reject H0 -> the observed data does NOT match the company's claimed 70% preference"
else:
    conclusion = "Fail to reject H0 -> the observed data is consistent with the claimed 70% preference"

print(f"Conclusion: {conclusion}")


Observed counts [Prefer, Do Not Prefer]: [18 12]
Expected counts based on claim (70% prefer): [21.  9.]

Chi-square statistic = 1.4286
p-value = 0.232
Conclusion: Fail to reject H0 -> the observed data is consistent with the claimed 70% preference


**Explanation:** A chi-square **goodness of fit** test checks whether observed data matches a claimed/expected distribution — here, whether the observed 18-out-of-30 preference rate is consistent with Flipkart's claim that 70% of users prefer the new product. The expected counts under the claim are 30 x 0.70 = 21 (prefer) and 30 x 0.30 = 9 (do not prefer). `scipy.stats.chisquare()` measures how large the gap is between observed (18/12) and expected (21/9) relative to random sampling variation, giving a p-value that tells us whether the actual observed preference (60%) is significantly different from the claimed 70%, or whether it's plausibly just chance variation from a sample of 30 users.